### ISSUE TITLE
QoS>0 messages stop prematurely in Windows.

### ISSUE THREAD

**Issue by ghaqua** [eclipse/mosquitto#1417]
Qos>0 messages in Windows are not published more than 20 messages. Specifying max_inflight_messages 0 in the broker's configuration file does not publish.

**Comment by karlp** [comment]
You're going to need to provide more detail, but I suspect that you're client applications are not calling mosquitto_loop or similar, so you simply hit max inflight and then it blocks waiting for the rest of the qos handshakes to complete.

**Comment by ralight** [comment]
You're missing a `mosquitto_loop_start()` call in the publishing client code, which means the network loop doesn't run. That means the client doesn't think it has received any `PUBACK` messages from the server, and so stops when it hits the limit of 20 inflight messages. It looks like you've found a bug, in that `mosquitto_max_inflight_messages_set()` isn't working properly.

**Comment by karlp** [comment]
I don't think anything is wrong with max in flight set(), they only call that in their subscriber call, and it's the publisher that's missing a loop_start/loop_forever that is simply unable to get any more messages out before it exits.

**Comment by ralight** [comment]
```
	int ii = mosquitto_max_inflight_messages_set(mosq, 0);
```

It's there in the pub code as well.

**Comment by ghaqua** [comment]
Thank you for your comment.
Below, I tried to change the publish part from (A) to (B), but the result was the same.
Is the program wrong?

(A)
	char	cPayload[4];
	for (int iCnt = 0; iCnt <= 500; iCnt++) {
		sprintf_s(cPayload, sizeof(cPayload), "%0d", iCnt);
		Sleep(1000);
		mosquitto_publish(mosq, &imid, "sndata", sizeof(cPayload), cPayload, 1, false);
	}

(B)

	mosquitto_loop_start(mosq);

	char	cPayload[4];
	for (int iCnt = 0; iCnt <= 500; iCnt++) {
		sprintf_s(cPayload, sizeof(cPayload), "%0d", iCnt);
		Sleep(1000);
		mosquitto_publish(mosq, &imid, "sndata", sizeof(cPayload), cPayload, 1, false);
	}

	mosquitto_loop_stop(mosq, false);

**Comment by ghaqua** [comment]
Thank you for the release and correction of 1.6.7.
Confirmation to set max_inflight_messages 0 with QoS=1 setting and publish more than 20 messages. However, after the client receives PUBREC when QoS=2 is made, it stops action and the message cannot be sent.

**Comment by ralight** [comment]
Have you fixed your code to have one of `mosquitto_loop_start()` or `mosquitto_loop_forever()`? You can confirm whether the problem is with your code by using mosquitto_sub to receive the messages. If that fails, it's a problem with the mosquitto code somewhere. If it works, I'm afraid it's your end.

**Comment by ghaqua** [comment]
Yes, I am using mosauitto _ loop _ forever (). As shown in the above sample program. I checked with mosquitto _ sub, but I still can't receive messages when QoS = 2.

**Comment by ralight** [comment]
Sorry, I got the programs the wrong way round. I meant that your publishing code above doesn't  have a loop call in it, so it will not process messages.

**Comment by ghaqua** [comment]
If QoS = 2 is specified in mosquitto _ sub and mosquitto _ pub, no message is sent.

mosquitto_sub -h localhost -t sndata -q 2
mosquitto_pub -h localhost -t sndata -q 2

**Comment by ralight** [comment]
Good to hear. So now what happens if you call `mosquitto_loop_start()` after `mosquitto_connect()` in your publishing client?

**Comment by ghaqua** [comment]
I don't know if I'm using it correctly, but calling mosquitto _ loop () after mosquitto _ connect () and calling mosquitto _ loop () after mosquitto _ publish () produced the expected results.

**Comment by ralight** [comment]
If your application does not want to do anything other than process incoming MQTT messages,  then the recommendation is to use `mosquitto_loop_forever()`.

If your application does need to do anything other than process incoming messages (for example you are publishing messages on a fixed interval), then the recommendation is to use `mosquitto_loop_start()` after calling `mosquitto_connect()`. That's it, nothing more is needed.

If you have other requirements, so as not wanting to use the extra thread created in `mosquitto_loop_start()`, then you may use `mosquitto_loop()`, but it must be called frequently. You should only do this if you are sure of what you are doing. You will need to implement reconnection logic, this is already covered for you by `mosquitto_loop_forever()` and `mosquitto_loop_start()`.

In short - use `mosquitto_loop_start()` for your publisher unless you have a compelling reason not to.

**Comment by ghaqua** [comment]
I understand.

However, calling mosquitto _ loop _ start () returned MOSQ _ ERR _ NOT _ SUPPORTED and was not available.

**Comment by ralight** [comment]
Threading support has been available in Windows for a while now, so I'm closing this.
